Math/NumberTheory/Zeta/Dirichlet.hs

```diff
@@ -0,0 +1,210 @@
+
+-- | The same sequence as @euler'@, but with type @[a]@ instead of @[Ratio a]@
+-- as the denominators in @euler'@ are always @1@.
+--
+-- >>> take 10 euler :: [Integer]
+-- [1, 0, -1, 0, 5, 0, -61, 0, 1385, 0]
+euler :: forall a . Integral a => [a]
+euler = map numerator euler'
+
+-- | Infinite zero-based list of the @n@-th order Euler polynomials evaluated at @1@.
+-- The algorithm used was derived from <http://www.emis.ams.org/journals/JIS/VOL4/CHEN/AlgBE2.pdf Algorithms for Bernoulli numbers and Euler numbers>
+-- by Kwang-Wu Chen, third formula of the Corollary in page 7.
+-- >>> take 10 eulerPolyAt1 :: [Rational]
+-- [1 % 1,1 % 2,0 % 1,(-1) % 4,0 % 1,1 % 2,0 % 1,(-17) % 8,0 % 1,31 % 2]
+eulerPolyAt1 :: forall a . Integral a => [Ratio a]
+eulerPolyAt1 = map (f . tail) (tail stirling2)
+  where
+    f = sum . zipWith4 (\sgn fact twos stir -> (sgn * fact * stir) % twos)
+                       (cycle [1, -1])
+                       factorial
+                       (iterate (2 *) 1)
+{-# SPECIALIZE eulerPolyAt1 :: [Ratio Int]     #-}
+{-# SPECIALIZE eulerPolyAt1 :: [Ratio Integer] #-}
+
+-- | Infinite sequence of exact values of Dirichlet beta-function at odd arguments, starting with @β(1)@.
+--
+-- > > approximateValue (betasOdd !! 25) :: Double
+-- > 0.9999999999999987
+--
+-- Using 'Data.Number.Fixed.Fixed':
+--
+-- > > approximateValue (betasOdd !! 25) :: Fixed Prec50
+-- > 0.99999999999999999999999960726927497384196726751694z
+--
+betasOdd :: [ExactPi]
+betasOdd = zipWith Exact [1, 3 ..] $ zipWith4 (\sgn denom eul twos -> sgn * (eul % (twos * denom)))
+                                              (cycle [1, -1])
+                                              (skipOdds factorial)
+                                              (skipOdds euler)
+                                              (iterate (4 *) 4)
+
+-- | @betasOdd@, but with @forall a . Floating a => a@ instead of @ExactPi@s.
+-- Used in @betasEven@.
+betasOdd' :: Floating a => [a]
+betasOdd' = map approximateValue betasOdd
+
+-- | Infinite sequence of approximate values of the Dirichlet @β@ function at
+-- positive even integer arguments, starting with @β(0)@.
+betasEven :: forall a. (Floating a, Ord a) => a -> [a]
+betasEven eps = (1 / 2) : bets
+  where
+    bets :: [a]
+    bets = zipWith3 (\r1 r2 r3 -> (r1 + (negate r2) + r3)) rhs1 rhs2 rhs3
+
+    evens = [0, 2 ..]
+
+    odds = [1, 3 ..]
+
+    -- [1!, 3!, 5!..]
+    factorial1AsInteger :: [Integer]
+    factorial1AsInteger = skipEvens factorial
+
+    -- [1!, 3!, 5!..]
+    factorial1 :: [a]
+    factorial1 = map fromInteger factorial1AsInteger
+
+    -- [1 / (2^1 * 1!), 1 / (2^3 * 3!), 1 /(2^5 * 5!), 1 / (2^7 * 7!) ..]
+    fracs :: [Rational]
+    fracs = zipWith (\pow fac -> 1 % (pow * fac)) factorial1AsInteger (iterate (4 *) 2)
+
+    -- First term of the right hand side of (12).
+    rhs1 = zipWith3 (\sgn piFrac lg -> sgn * (lg * approximateValue piFrac)) (cycle [1, -1])
```

Comment:
Can we just use `log 2` instead of zipping with an infinite stream of `repeat (log 2)`?

`approximateValue` does nothing fancy: `approximateValue (Exact z q) = (pi ^^ z) * (fromRational q)`. So there is no benefit to save something as `ExactPi` just to be directly consumed by `approximateValue`; one can immediately work with `pi :: Floating a => a`.